dynamodb/scan.go

```diff
@@ -6,46 +6,63 @@ import (
 	simplejson "github.com/bitly/go-simplejson"
 )
 
-func (t *Table) FetchResults(query *Query) ([]map[string]*Attribute, error) {
+func (t *Table) FetchResults(startKey *Key, query *Query) ([]map[string]*Attribute, *Key, error) {
+	if startKey != nil {
+		query.AddStartKey(t, startKey)
+	}
 	jsonResponse, err := t.Server.queryServer(target("Scan"), query)
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
-
 	json, err := simplejson.NewJson(jsonResponse)
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
 
 	itemCount, err := json.Get("Count").Int()
 	if err != nil {
 		message := fmt.Sprintf("Unexpected response %s", jsonResponse)
-		return nil, errors.New(message)
+		return nil, nil, errors.New(message)
 	}
 
-	results := make([]map[string]*Attribute, itemCount)
+	var lastKey *Key
+	if marker, ok := json.CheckGet("LastEvaluatedKey"); ok {
+		keymap, err := marker.Map()
+		if err != nil {
+			return nil, nil, fmt.Errorf("Unexpected LastEvaluatedKey in response %s\n", jsonResponse)
+		}
+		lastKey = &Key{}
+		hashmap := keymap[t.Key.KeyAttribute.Name].(map[string]interface{})
+		lastKey.HashKey = hashmap[t.Key.KeyAttribute.Type].(string)
+		if t.Key.HasRange() {
+			rangemap := keymap[t.Key.RangeAttribute.Name].(map[string]interface{})
+			lastKey.RangeKey = rangemap[t.Key.RangeAttribute.Type].(string)
+		}
+	}
 
+	results := make([]map[string]*Attribute, itemCount)
 	for i, _ := range results {
 		item, err := json.Get("Items").GetIndex(i).Map()
 		if err != nil {
 			message := fmt.Sprintf("Unexpected response %s", jsonResponse)
-			return nil, errors.New(message)
+			return nil, lastKey, errors.New(message)
 		}
 		results[i] = parseAttributes(item)
 	}
-	return results, nil
+	return results, lastKey, nil
 
 }
 
-func (t *Table) Scan(attributeComparisons []AttributeComparison) ([]map[string]*Attribute, error) {
```

Comment:
Maybe we should keep a version that doesn't take a start and add the new method as `ScanWithStart` so we can ignore upstream tests.